adapters/handlers/grpc/parse_request.go

```diff
@@ -253,20 +293,44 @@ func extractDataType(scheme schema.Schema, operator filters.Operator, classname
 		dataType = schema.DataTypeBoolean
 	} else if len(on) > 1 {
 		propToCheck := on[len(on)-1]
+		_, isPropLengthFilter := schema.IsPropertyLength(propToCheck, 0)
+		if isPropLengthFilter {
+			return schema.DataTypeInt, nil
+		}
+
 		classOfProp := on[len(on)-2]
 		prop, err := scheme.GetProperty(schema.ClassName(classOfProp), schema.PropertyName(propToCheck))
 		if err != nil {
 			return dataType, err
 		}
 		return schema.DataType(prop.DataType[0]), nil
 	} else {
-		prop, err := scheme.GetProperty(schema.ClassName(classname), schema.PropertyName(on[0]))
+		propToCheck := on[0]
+		_, isPropLengthFilter := schema.IsPropertyLength(propToCheck, 0)
+		if isPropLengthFilter {
+			return schema.DataTypeInt, nil
+		}
+
+		prop, err := scheme.GetProperty(schema.ClassName(classname), schema.PropertyName(propToCheck))
 		if err != nil {
 			return dataType, err
 		}
 		dataType = schema.DataType(prop.DataType[0])
 	}
-	return dataType, nil
+
+	if operator == filters.ContainsAll || operator == filters.ContainsAny {
+		if dataType == schema.DataTypeText {
+			return dataType, nil
+		}
+		baseType, isArray := schema.IsArrayType(dataType)
+		if !isArray {
+			return baseType, fmt.Errorf("operator %v can only be used on text and array fields, datatype is %v", operator, dataType)
```

Comment:
> contains is not designed to be used with texts (though it might be extended in the future, same as len() works with array and text props).

hmmm, but it works 😅 See here:https://github.com/weaviate/weaviate-python-client/blob/53ee009bccce61e57c3bfa18c471601c9c73d5eb/integration/test_collection.py#L1252

> while for ContainsAny should return some results (value is equal to X OR Y)
Thanks, will add that!

Reply:
documentation also says that it works:
https://weaviate.io/developers/weaviate/api/graphql/filters#containsany--containsall

> The ContainsAny and ContainsAll operators treat texts as an array. The text is split into an array of tokens based on the chosen tokenization scheme, and the search is performed on that array.

Reply:
> Contains works as well with non-array types

fixed